### DISCUSSION TITLE
How to set window property?

### DISCUSSION THREAD

**Discussion by ervinpopescu** [qtile/qtile#4436] [Q&A]
Trying:
```py
@hook.subscribe.client_focus
@hook.subscribe.client_new
def resize_and_move_client(client: Window):
    client.window.set_property("_NET_WM_WINDOW_TYPE", "_NET_WM_WINDOW_TYPE_DESKTOP")
```
fails with:
```py
2023-09-03 15:55:13,011 ERROR libqtile hook.py:fire():L903 Error in hook client_focus
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/libqtile/hook.py", line 901, in fire
    i(*args, **kwargs)
  File "/home/ervin/.config/qtile/modules/hooks.py", line 122, in resize_and_move_client
    client.window.set_property("_NET_WM_WINDOW_TYPE", "_NET_WM_WINDOW_TYPE_DESKTOP")
  File "/usr/lib/python3.11/site-packages/libqtile/backend/x11/window.py", line 358, in set_property
    ).check()
      ^^^^^^^
  File "/home/ervin/.local/lib/python3.11/site-packages/xcffib/__init__.py", line 337, in check
    self.conn.request_check(self.sequence)
  File "/home/ervin/.local/lib/python3.11/site-packages/xcffib/__init__.py", line 579, in wrapper
    return f(*args)
           ^^^^^^^^
  File "/home/ervin/.local/lib/python3.11/site-packages/xcffib/__init__.py", line 689, in request_check
    self._process_error(err)
  File "/home/ervin/.local/lib/python3.11/site-packages/xcffib/__init__.py", line 657, in _process_error
    raise error(buf)
xcffib.xproto.LengthError
```

**Comment by elParaguayo** [comment]
I think you've got to pass the atom value, rather than a string. Try:
```python
@hook.subscribe.client_focus
@hook.subscribe.client_new
def resize_and_move_client(client: Window):
    win = client.window
    atom = win.conn.atoms[ "_NET_WM_WINDOW_TYPE_DESKTOP"]
    win.set_property("_NET_WM_WINDOW_TYPE", [atom])
```
You should probably be checking whether this value is already set and whether you want to do it on every window...

**Reply by ervinpopescu** [reply]
No no, I'm not setting it on every window :) This is the complete function:
```py
with open(os.path.join(config_path, "json", "window_rules.json"), "r") as f:
    rules: dict = json5.loads(f.read())

@hook.subscribe.client_focus
@hook.subscribe.client_new
def resize_and_move_client(client: Window):
    try:
        wm_class = client.window.get_wm_class()[0]
    except Exception:
        wm_class = None
    try:
        role = client.get_wm_role()
    except Exception:
        role = None

    for key, win in rules.items():
        if key in [wm_class, role]:
            if "set_position_floating" in win and key == "gsimplecal":
                client.set_position_floating(
                    x=2880 - win["w"] - settings["margin_size"] - 5,
                    y=settings["bar_height"] + 2 * settings["margin_size"],
                )

            if "set_size_floating" in win:
                if key == "blueman-manager":
                    sleep(3)
                client.set_size_floating(w=win["w"], h=win["h"])

            if "toggle_floating" in win:
                client.toggle_floating()

            if "center" in win:
                client.center()

            if "keep_above" in win:
                client.keep_above()

    if "Musializer" in [wm_class, role]:
        win = client.window
        atom = win.conn.atoms["_NET_WM_WINDOW_TYPE_DESKTOP"]
        win.set_property("_NET_WM_WINDOW_TYPE", [atom])
```
This does get set correctly, but for some reason Qtile does not react to it:
```xprop
_NET_WM_STATE(ATOM) =
_NET_FRAME_EXTENTS(CARDINAL) = 5, 5, 5, 5
ZOOM(UTF8_STRING) = "1.0"
_NET_WM_DESKTOP(CARDINAL) = 0
WM_STATE(WM_STATE):
                window state: Normal
                icon window: 0x0
_VARIABLE_REFRESH(CARDINAL) = 1
_NET_WM_ICON_NAME(UTF8_STRING) = "Musializer"
_NET_WM_NAME(UTF8_STRING) = "Musializer"
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "arch"
WM_ICON_NAME(STRING) = "Musializer"
WM_NAME(STRING) = "Musializer"
XdndAware(ATOM) = BITMAP
WM_CLASS(STRING) = "Musializer", "Musializer"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                window gravity: Static
WM_HINTS(WM_HINTS):
                Initial state is Normal State.
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_DESKTOP
_NET_WM_PID(CARDINAL) = 975631
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING
```
How can I make sure that the window gets a desktop type and stays below every window? And will it be under Internal windows too?

**Reply by elParaguayo** [reply]
Well, you should really just be running `client.keep_below(enable=True)` rather than doing this as the window isn't a desktop. If that doesn't work, try adding `client.change_layer()` to the end of your code to force the restacking.

**Reply by ervinpopescu** [reply]
Yeah, I just completely gave up on it. Could you please provide any resources that will help me understand X11 better? I'm not exactly a Wayland enthusiast, so I won't switch over any time soon :)

**Reply by elParaguayo** [reply]
There's no real resource for X11 on this (as far as I can tell) other than EWMH spec which we've tried to implement here. Our docs on this are here: https://docs.qtile.org/en/latest/manual/stacking.html

**Reply by elParaguayo** [reply]
If stacking isn't working properly then please post an issue and I'll take a look when I can.

**Reply by ervinpopescu** [reply]
I don't want to open an issue, as I'm unsure how this is supposed to work internally. That is why I asked you if you have found any sort of X11 reference that could help me understand the whole server-client communication and other stuff better. This is most likely intended behaviour, as in "The application/window should set itself as _NET_WM_WINDOW_TYPE_DESKTOP".

**Reply by elParaguayo** [reply]
You probably want this: https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html

**Reply by ervinpopescu** [reply]
> This SHOULD be set by the Client before mapping to a list of atoms indicating the functional type of the window. This property SHOULD be used by the window manager in determining the decoration, stacking position and other behavior of the window. The Client SHOULD specify window types in order of preference (the first being most preferable) but MUST include at least one of the basic window type atoms from the list below.

I don't remember reading this, but I'm glad (😭) I was right. Anyhow, this is weird and will look into it further. The spec is also clunky and it makes sense that people wanted more (and got Wayland), just saying.
